Yum resource fix for non-existent repos and repo info

adamleff opened the pull request:
If a repo did not exist, running matchers against it (such as `exist`)
were failing due to a bug in `#to_s` when fetching the repo name. The
`info` method would return nil and we'd still try to treat it as a hash.

This change ensures that info is always a hash, possibly empty if the
repo doesn't exist, and uses the repo name provided by the user rather
than shortening it to be consistent with our other resources which don't
manipulate the user input in the formatter.

Also added a method_missing to allow users to interrogate repo options,
such as baseurl or gpgcheck.

Fixes #1553

adamleff committed:
Yum resource fix for non-existent repos and repo info

If a repo did not exist, running matchers against it (such as `exist`)
were failing due to a bug in `#to_s` when fetching the repo name. The
`info` method would return nil and we'd still try to treat it as a hash.

This change ensures that info is always a hash, possibly empty if the
repo doesn't exist, and uses the repo name provided by the user rather
than shortening it to be consistent with our other resources which don't
manipulate the user input in the formatter.

Also added a method_missing to allow users to interrogate repo options,
such as baseurl or gpgcheck.

Signed-off-by: Adam Leff <adam@leff.co>

shoekstra commented on lib/resources/yum.rb:
```diff
@@ -33,7 +32,7 @@
   class Yum < Inspec.resource(1)
     name 'yum'
-    desc 'Use the yum InSpec audit resource to test packages in the Yum repository.'
+    desc 'Use the yum InSpec audit resource to the configuration of Yum repositories.'
```
This is missing the word "test" and should be:

```desc 'Use the yum InSpec audit resource to test the configuration of Yum repositories.'```

Sorry for being pedantic :)

adamleff replied:
Pedants welcome. I'm the club president.

chris-rock commented on lib/resources/yum.rb:
```diff
@@ -121,22 +120,25 @@ def shortname(id)
       info['status'] == 'enabled'
     end
 
+    def method_missing(name)
```
We try to avoid method_missing whenever possible. Do we know all possible properties? Maybe we use something like:

```
    %w{
      baseurl
    }.each do |m|
      define_method m.to_sym do |*args|
        info[m.to_s]
      end
    end
```

chris-rock commented on lib/resources/yum.rb:
```diff
@@ -22,6 +20,7 @@
 # describe yum.repo('epel') do
 #   it { should exist }
 #   it { should be_enabled }
+#   its('baseurl') { should include 'mycompany.biz' }
```
should this be `baseurl` or ruby style `base_url`?

adamleff replied:
The key in the yum output and in the `*.repo` files is `baseurl` so I vote for keeping it consistent.

chris-rock commented on lib/resources/yum.rb:
```diff
@@ -121,22 +120,25 @@ def shortname(id)
+
     def to_s
-      "YumRepo #{shortname(info['id'])}"
+      "YumRepo #{@reponame}"
```
Thank you 👍

adamleff committed:
Remove method_missing, provide methods for repo metadata

Instead of method_missing, methods for each output item from
`yum repolist` are provided.

Signed-off-by: Adam Leff <adam@leff.co>

arlimus commented on lib/resources/yum.rb:
```diff
@@ -121,22 +120,37 @@ def shortname(id)
     def enabled?
-      repo = info
-      return false if repo.nil?
+      return false unless exist?
```
Nice simplification

arlimus reviewed (approved):
Great update and more usefull methods here, kudos @adamleff
